### PR TITLE
feat(dds): do not opaquely encode handles in ops

### PR DESCRIPTION
**Pull request by connorskees**
Update DDSes to not opaquely encode handles as strings, but rather pass around actual objects.

[AB#6382](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/6382)

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 1]
```diff
@@ -451,7 +457,7 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 				this.reSubmit(content, localOpMetadata);
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:457]: Should this also decode handles, and then the default implementation re-encodes them?

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 2]
```diff
@@ -102,25 +110,72 @@ describe("DDS Handle Encoding", () => {
 			},
 			[handle.absolutePath] /* expectedHandles */,
 		),
+		createTestCase(
+			new SharedStringFactory(),
+			(dds) => {
+				dds.insertMarker(0, ReferenceType.Simple, { marker: handle });
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:116]: Any other APIs on SharedString where someone could stash a handle?

Comment (markfields) [reply packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:116]: Like `annotateMarker` or `insertText`?  I don't know anything about how it's implemented so maybe it's completely obvious that the one test case is enough. But if not, maybe good to cover a few scenarios here.  I figure you can use this one test case and just put more handles in `expectedHandles`

Comment (connorskees) [reply packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:116]: They're all sort of the same; the handles would get stored in the same place. Let me think a bit about if there's anywhere else you could theoretically put a handle.

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 3]
```diff
@@ -102,25 +110,72 @@ describe("DDS Handle Encoding", () => {
...
+				dds.setCell(0, 0, handle);
+			},
+			[handle.absolutePath] /* expectedHandles */,
+		),
+		createTestCase(
+			new TreeFactory({}),
+			(dds: ITree) => {
+				const builder = new SchemaFactory("test");
+				class Bar extends builder.object("bar", {
+					h: builder.optional(builder.handle),
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:135]: Would any other structural pieces represent distinct test cases, or is this sufficient?

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 4]
```diff
@@ -102,25 +110,72 @@ describe("DDS Handle Encoding", () => {
...
 			},
-			[] /* expectedHandles */,
+			[handle.absolutePath] /* expectedHandles */,
+		),
+		createTestCase(
+			new ConsensusQueueFactory(),
+			(dds) => {
+				dds.add(handle).catch(() => {});
+			},
+			[handle.absolutePath] /* expectedHandles */,
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:160]: I thought this one wasn't going to work.  I haven't looked at the code yet though...

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 5]
```diff
@@ -102,25 +110,72 @@ describe("DDS Handle Encoding", () => {
...
+		),
+		createTestCase(
+			new CellFactory(),
+			(dds) => {
+				dds.set(handle);
+			},
+			[handle.absolutePath] /* expectedHandles */,
 		),
+		// can't have handles: counter, task manager,
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:169]: nit
```suggestion
		// API doesn't support handles: counter, task manager,
```

File: packages/framework/attributor/src/test/attribution/sharedString.attribution.spec.ts [hunk 6]
```diff
@@ -662,7 +662,7 @@ describe("SharedString Attribution", () => {
 			},
 		];
 
-		it.skip("Generate a new document", async () => {
+		it.only("Generate a new document", async () => {
```
Comment (markfields) [inline packages/framework/attributor/src/test/attribution/sharedString.attribution.spec.ts:665]: oops

File: packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json [hunk 7]
```diff
@@ -10,7 +10,7 @@
...
...1\",\"prop3\":\"5KoWp\"},\"type\":2},{\"pos1\":31,\"pos2\":40,\"props\":{\"prop2\":\"dHWn1\",\"prop3\":\"5KoWp\"},\"type\":2}],\"type\":3},\"referenceSequenceNumber\":81,\"type\":\"op\",\"sequenceNumber\":82,\"minimumSequenceNumber\":76},{\"clientId\":\"d3fdea22-7906-496f-9fc5-0bf988954452\",\"clientSequenceNumber\":29,\"contents\":{\"pos1\":31,\"seg\":\"yc5Bz\",\"type\":0},\"referenceSequenceNumber\":82,\"type\":\"op\",\"sequenceNumber\":83,\"minimumSequenceNumber\":76},{\"clientId\":\"d3fdea22-7906-496f-9fc5-0bf988954452\",\"clientSequenceNumber\":30,\"contents\":{\"pos1\":36,\"seg\":\"mQny\",\"type\":0},\"referenceSequenceNumber\":83,\"type\":\"op\",\"sequenceNumber\":84,\"minimumSequenceNumber\":76},{\"clientId\":\"25adaa92-9220-49e2-a852-3c7e97ac9d1c\",\"clientSequenceNumber\":24,\"contents\":{\"ops\":[],\"type\":3},\"referenceSequenceNumber\":84,\"type\":\"op\",\"sequenceNumber\":85,\"minimumSequenceNumber\":76},{\"clientId\":\"a2df62bb-461f-4c2e-a3ad-fc5df463c464\",\"clientSequenceNumber\":29,\"contents\":{\"pos1\":4,\"seg\":\"x2NAdmR\",\"type\":0},\"referenceSequenceNumber\":85,\"type\":\"op\",\"sequenceNumber\":86,\"minimumSequenceNumber\":76,\"timestamp\":1665600040000}]"
```
Comment (markfields) [inline packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json:13]: Is this related to this PR? All I see when skimming the diff is a timestamp.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 8]
```diff
@@ -345,6 +346,8 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 	 */
 	protected abstract onDisconnect();
 
+	protected abstract get serializer(): IFluidSerializer;
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:349]: doc comment please

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 9]
```diff
@@ -345,6 +346,8 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
...
+
 	/**
 	 * Submits a message by the local client to the runtime.
 	 * @param content - Content of the message
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:353]: Mention something like this?

```suggestion
	 * @param content - Content of the message
	 * Note: handles contained in the message object should not be encoded in any way
```

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (markfields) [reply packages/dds/shared-object-base/src/sharedObject.ts:457]: Ping @connorskees, @anthony-murphy

Comment (markfields) [reply packages/dds/shared-object-base/src/sharedObject.ts:457]: And, would be excellent to get test coverage of roundtripping through these scenarios (resubmit and stashed op, maybe something with rollback)

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 10]
```diff
@@ -442,7 +448,7 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 				local: boolean,
 				localOpMetadata: unknown,
 			) => {
-				this.process(message, local, localOpMetadata);
+				this.process(parseHandles(message, this.serializer), local, localOpMetadata);
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:451]: About optimizing perf impact*:

`parseHandles` does a roundtrip through string, I wonder if we should update it to use `IFluidSerializer.decode` which does its own object walk?

Also, `parseHandles` only needs to be done on `message.contents`

\* If this PR was simply moving `parseHandles` calls from DDS implementations to this one place, it could be a separate follow-up.  But I see at least one place (`sharedTreeCore.ts`) where previously it was using `decode` but now will use the (presumably slower) `parseHandles`.

**Review by markfields — COMMENTED** [review]
I got through the tests and the SharedObject code, glanced at some of the DDS changes.  I'll come back tomorrow to finish.

File: packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json (hunk 7 above)
Comment (connorskees) [reply packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json:13]: It appears caused by this PR, but as you've mentioned is largely harmless and just adds the timestamp field to the snapshot. It's not clear to me why this occurs.

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts (hunk 3 above)
Comment (connorskees) [reply packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:135]: I think this combined with the handle tests in the shared-tree package directly give sufficient coverage.

File: packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json (hunk 7 above)
Comment (markfields) [reply packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json:13]: Feels like the kind of thing that's worth understanding why. As we both know, there's relevant code here that roundtrips through JSON string (or not), maybe the PR changes the exact logic there in a way that triggers serialization to change somehow.

cc @Abe27342 in case he has comments on the purpose of these snapshots and why we might care about an apparently harmless change.

File: packages/dds/sequence/src/sequence.ts [hunk 11]
```diff
@@ -459,9 +458,9 @@ export abstract class SharedSegmentSequence<T extends ISegment>
 		// local ops until loading is complete, and then
 		// they will be present
 		if (!this.loadedDeferred.isCompleted) {
-			this.loadedDeferredOutgoingOps.push(metadata ? [translated, metadata] : translated);
+			this.loadedDeferredOutgoingOps.push(metadata ? [message, metadata] : (message as any));
```
Comment (markfields) [inline packages/dds/sequence/src/sequence.ts:461]: Side note - The code that reads from this `loadedDeferredOutgoingOps` doesn't seem to account for the case where `metadata === undefined`...

@anthony-murphy - do you see what I'm talking about? AFAIK, the destructuring on line 873 (the `for` line) would throw in this case.

File: packages/dds/sequence/src/sequence.ts [hunk 12]
```diff
@@ -684,7 +683,7 @@ export abstract class SharedSegmentSequence<T extends ISegment>
 								})}`,
 							);
 						}
-						this.processMergeTreeMsg(m);
+						this.processMergeTreeMsg(parseHandles(m, this.serializer));
```
Comment (markfields) [inline packages/dds/sequence/src/sequence.ts:686]: I was wondering what about the other call site - but I see that `processCore` takes in an already-decoded object.  Since we don't have expressive types here to distinguish those cases, I'd suggest making it clear in the params comment for each function at this seam whether the message is expected to have handle encoded or decoded.

Comment (markfields) [reply packages/dds/sequence/src/sequence.ts:686]: e.g. checkout `loadCatchupOps` in `snapshotLoader.ts`. That's the function that originates the list of messages being iterated over here, from my reading of the code. There's a comment there saying:

> The decoded messages, but handles aren't parsed.  Matches the format that will be passed in SharedObject.processCore

So (A) that's a helpful comment and (B) consider moving this `parseHandles` over there, to maintain the symmetry ("matches the format that will be passed in SharedObject.processCore"). (or if you don't then update that comment to say it doesn't match)

File: packages/dds/sequence/src/defaultMap.ts [hunk 13]
```diff
@@ -286,7 +281,7 @@ export class DefaultMap<T> {
 	}
 
 	public populate(json: string): void {
-		this.populateFromSerializable(JSON.parse(json) as IMapDataObjectSerializable);
+		this.populateFromSerializable(this.serializer.parse(json) as IMapDataObjectSerializable);
```
Comment (markfields) [inline packages/dds/sequence/src/defaultMap.ts:284]: This is surprising, based on the name `populateFromSerializable`. The object passed in there will have handles in non-serializable form

File: packages/dds/sequence/src/defaultMap.ts [hunk 14]
```diff
@@ -399,7 +393,7 @@ export class DefaultMap<T> {
 				const localValue = this.data.get(op.key) ?? this.createCore(op.key, local);
 				const handler = localValue.getOpHandler(op.value.opName);
 				const previousValue = localValue.value;
-				const translatedValue = parseHandles(op.value.value, this.serializer);
+				const translatedValue = op.value.value as any;
```
Comment (markfields) [inline packages/dds/sequence/src/defaultMap.ts:396]: All this ValueType stuff is new to me - can you just tell me where this `parseHandles` moves to in the PR? I traced through to see that the `message` param has already been parsed, but I don't know where `op` comes from (and trying to move through the PR)

Comment (markfields) [reply packages/dds/sequence/src/defaultMap.ts:396]: Ahh I spoke too soon :) The `emit` function defined below in `makeMapValueOpEmitter` no longer encodes the handles.  I'm assuming that this inner nugget will have its handles encoded when the whole op goes to `SharedObject.submitLocalMessage`?  And this code path has test coverage in the new functional test cases you're adding?

Comment (markfields) [reply packages/dds/sequence/src/defaultMap.ts:396]: > this code path has test coverage

I could be wrong but I put a breakpoint in the `emit` function below and ran the tests and it didn't get hit.  Please double-check that this has test coverage (though for the record I do agree the change looks right)

File: packages/dds/register-collection/src/consensusRegisterCollection.ts [hunk 15]
```diff
@@ -67,7 +67,7 @@ interface IRegisterOperation {
 }
 
 /**
- * IRegisterOperation format in versions \< 0.17
+ * IRegisterOperation format in versions \< 0.17 and \>8.0.0
```
Comment (markfields) [inline packages/dds/register-collection/src/consensusRegisterCollection.ts:70]:
```suggestion
 * IRegisterOperation format in versions \< 0.17 and \>2.0.0-internal.8.0.0
```

File: packages/dds/register-collection/src/consensusRegisterCollection.ts [hunk 16]
```diff
@@ -142,18 +142,19 @@ export class ConsensusRegisterCollection<T>
 	 * @returns Promise<true> if write was non-concurrent
 	 */
 	public async write(key: string, value: T): Promise<boolean> {
-		const serializedValue = this.stringify(value, this.serializer);
-
 		if (!this.isAttached()) {
 			// JSON-roundtrip value for local writes to match the behavior of going through the wire
```
Comment (markfields) [inline packages/dds/register-collection/src/consensusRegisterCollection.ts:146]:
```suggestion
```

File: packages/dds/register-collection/src/consensusRegisterCollection.ts [hunk 17]
```diff
@@ -142,18 +142,19 @@ export class ConsensusRegisterCollection<T>
...
-			this.processInboundWrite(key, this.parse(serializedValue, this.serializer), 0, 0, true);
+			this.processInboundWrite(key, value, 0, 0, true);
 			return true;
 		}
 
-		const message: IRegisterOperation = {
+		const message: IRegisterOperationOld<T> = {
```
Comment (markfields) [inline packages/dds/register-collection/src/consensusRegisterCollection.ts:151]: Rename this type to not say `Old` anymore?

File: packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json (hunk 7 above)
Comment (Abe27342) [reply packages/framework/attributor/src/test/attribution/documents/default/attributor-lz4-and-delta-snap.json:13]: It's not really concerning to me. Real SharedString snapshots will have timestamps since they just stick the `ISequencedDocumentMessage` in their format (which is also why we have some back-compat code dealing with "term"). Our mocks--which this test suite use--don't set up timestamps by default. Since attribution cares about timestamps, it sets them up on its own in a bit of a kludgy way, see below:

image

This kludge ended up not carrying over into the catchup ops part of the snapshot before, probably due to some object copies that you've now removed, and it's plausible that they do now. This test suite is pretty much only concerned with attribution snapshot format anyway, which you haven't made changes to.

File: packages/dds/register-collection/src/consensusRegisterCollection.ts (hunk 17 above)
Comment (markfields) [reply packages/dds/register-collection/src/consensusRegisterCollection.ts:151]: And likewise rename `incomingOpMatchesCurrentFormat`, it's backwards now according to that name

File: packages/dds/register-collection/src/consensusRegisterCollection.ts [hunk 18]
```diff
@@ -67,7 +67,7 @@ interface IRegisterOperation {
 }
```
Comment (markfields) [inline packages/dds/register-collection/src/consensusRegisterCollection.ts:67]: So the "current" type (as of before this PR) has `refSeq: number | undefined`. The "old" type (now used again in this PR) has `refSeq: number`.  Do we need to account for the `undefined` case or not?

Also, that comment about the semantics of `refSeq` should be brought over too.

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 19]
```diff
@@ -102,25 +110,72 @@ describe("DDS Handle Encoding", () => {
...
+				dds.write("whatever", handle).catch(() => {});
+			},
+			[handle.absolutePath] /* expectedHandles */,
+		),
+		createTestCase(
+			new ConsensusQueueFactory(),
+			(dds) => {
+				dds.add(handle).catch(() => {});
 			},
+			// todo: AB#7149 this DDS does not currently support storing handles
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:160]: Personal preference: I've seen too many 5-year-old `// todo:` comments, I don't think they're an effective way to keep track of work.  Better just to say plainly "This DDS does not...".

If someone's curious about context, they can do git blame and see this PR.  Hopefully your description will mention something about this and other interesting cases.

Also I don't think we want AB links in the source code.

File: packages/dds/map/src/mapKernel.ts [hunk 20]
```diff
@@ -184,7 +184,7 @@ export class MapKernel {
 		private readonly isAttached: () => boolean,
 		private readonly eventEmitter: TypedEventEmitter<ISharedMapEvents>,
 	) {
-		this.localValueMaker = new LocalValueMaker(serializer);
+		this.localValueMaker = new LocalValueMaker();
```
Comment (markfields) [inline packages/dds/map/src/mapKernel.ts:187]: Why moving this from constructor param? I would expect the opposite - that we would get rid of `private readonly serializer` from `MapKernel`.  Looking at its usages, it's used in `set`, I think that's a `makeSerializable` that you want to remove.  Take a look, I could be missing something.

File: packages/dds/map/src/localValues.ts [hunk 21]
```diff
@@ -103,29 +103,34 @@ export class PlainLocalValue implements ILocalValue {
...
 		if (serializable.type === ValueType[ValueType.Shared]) {
 			serializable.type = ValueType[ValueType.Plain];
 			const handle: ISerializedHandle = {
 				type: "__fluid_handle__",
 				url: serializable.value as string,
 			};
-			serializable.value = handle;
+			// NOTE: here we require the use of `parseHandles` because the roundtrip
+			// through a string is necessary to resolve the absolute path of
+			// legacy handles (`ValueType.Shared`)
```
Comment (markfields) [inline packages/dds/map/src/localValues.ts:128]: I don't understand why the roundtrip through string is necessary. In this case, the incoming object `serializable` is a legacy format... as I think about it I don't understand how this can be. Or maybe worse - this may represent a gap in parsing handles in SharedObject. i.e. in your PR this is unreachable code (and this legacy handle wouldn't be parsed by SharedOject).

Aside from this case, my assumption is all the objects in this package have real handles. Because we are now parsing incoming handles in SharedObject.  BuT this is apparently a case where the generic `parseHandle` logic isn't sufficient.

So we may need to inject this backcompat code into the SharedObject thing.  i.e. DDSes get to provide a callback in case there's custom handle conversion logic required.

Comment (markfields) [reply packages/dds/map/src/localValues.ts:128]: Ok Connor set me straight offline:

> this legacy handle wouldn't be parsed by SharedOject

This is true, and hence we will still get to this codepath. So no, "this is unreachable code" is not correct.

And he explained how in this legacy case the handle path may be relative, and depends on some patch-up code in serializer's `decode` logic to switch to absolute path.

File: packages/dds/map/src/directory.ts [hunk 22]
```diff
@@ -826,7 +831,8 @@ export class SharedDirectory
 					const localValue = this.makeLocal(
 						key,
 						currentSubDir.absolutePath,
-						serializable,
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+						parseHandles(serializable, this.serializer),
```
Comment (markfields) [inline packages/dds/map/src/directory.ts:835]: Maybe instead, replace `readAndParse` (which does `JSON.parse`) with `readBlob` + `serializer.parse`? That's preferable because then it's consistent throughout this code - handles are not encoded. Check out `loadCore` in other DDSes, I glanced at Matrix and CRC for reference.

File: packages/dds/map/src/map.ts [hunk 23]
```diff
@@ -334,7 +334,10 @@ export class SharedMap extends SharedObject<ISharedMapEvents> implements IShared
 		const json = await readAndParse<object>(storage, snapshotFileName);
```
Comment (markfields) [inline packages/dds/map/src/map.ts:334]: See https://github.com/microsoft/FluidFramework/pull/19242#discussion_r1484715046

I think it'd be better to do `readBlob` then `this.serializer.parse`, to be consistent that handles are always unencoded in this code.

File: packages/dds/map/api-report/map.api.md [hunk 24]
```diff
@@ -191,9 +191,9 @@ export interface IValueChanged {
 
 // @alpha
 export class LocalValueMaker {
-    constructor(serializer: IFluidSerializer);
+    constructor();
     fromInMemory(value: unknown): ILocalValue;
-    fromSerializable(serializable: ISerializableValue): ILocalValue;
+    fromSerializable(serializable: ISerializableValue, serializer: IFluidSerializer, bind: IFluidHandle): ILocalValue;
```
Comment (markfields) [inline packages/dds/map/api-report/map.api.md:196]: Technically a breaking change to `@alpha`. But not caught by type tests due to a known issue. And thankfully in practice this class isn't used externally (and shouldn't be).

So it's fine, but awkward.

File: packages/dds/cell/src/cell.ts [hunk 25]
```diff
@@ -227,7 +227,8 @@ export class SharedCell<T = any>
 	protected async loadCore(storage: IChannelStorageService): Promise<void> {
 		const content = await readAndParse<ICellValue>(storage, snapshotFileName);
```
Comment (markfields) [inline packages/dds/cell/src/cell.ts:228]: Same comment as directory - probably better to just use `readBlob` + `this.serializer.parse`

File: packages/dds/cell/src/cell.ts [hunk 26]
```diff
@@ -324,8 +325,7 @@ export class SharedCell<T = any>
 	}
 
 	private decode(cellValue: ICellValue): Serializable<T> {
```
Comment (markfields) [inline packages/dds/cell/src/cell.ts:327]: Either inline this or rename it

File: experimental/dds/tree/src/SharedTree.ts [hunk 27]
```diff
@@ -1401,7 +1400,7 @@ export class SharedTree extends SharedObject<ISharedTreeEvents> implements NodeI
 	}
 
 	private serializeEdit<TChange>(preparedEdit: Edit<TChange>): Edit<TChange> {
```
Comment (markfields) [inline experimental/dds/tree/src/SharedTree.ts:1402]: @CraigMacomber, @noencke - FYI the "semi-serializable" state is going away.  Handles will remain unencoded all throughout the DDS implementation.  `SharedObject.submitLocalMessage` is becoming the chokepoint for making the handles serializable.  And eventually that will move up to the ContainerRuntime.

I think we should probably open a small task to go clean up all the places that use this callback, etc.

Comment (markfields) [reply experimental/dds/tree/src/SharedTree.ts:1402]: @connorskees - Until then, let's inline this (i.e. use `(x) => x`), or at least rename this private function.

File: packages/dds/map/api-report/map.api.md (hunk 24 above)
Comment (markfields) [reply packages/dds/map/api-report/map.api.md:196]: > So it's fine

To clarify why it's fine - breaking changes are allowed in `main` now - but when action is requied on the consumer's side to digest the breaking change, we try to release the new alternative 1 release ahead of time. This does not apply here.

File: experimental/dds/attributable-map/src/localValues.ts [hunk 28]
```diff
@@ -134,9 +128,7 @@ export class LocalValueMaker {
 			serializable.value = handle;
 		}
 
-		const translatedValue: unknown = parseHandles(serializable.value, this.serializer);
```
Comment (markfields) [inline experimental/dds/attributable-map/src/localValues.ts:137]: I assume you need to do the same back-compat affordance here that you did in the other LocalValueMaker?

**Comment by rajatch-ff** [comment]
@connorskees Is the PR ready for review? If yes, do you wanna update the PR Description to share the context and describe the changes? Would be helpful to read before heading over to the file.

**Comment by rajatch-ff** [comment]
Thanks for updating @connorskees . A little more context would help someone unfamiliar with the area. For instance, what does opaquely encode handles as string means? What's the problem with that? And why are we moving away from it and passing around actual objects now. Who does this whole work help and how?

---
In reply to: [1936543215](https://github.com/microsoft/FluidFramework/pull/19242#issuecomment-1936543215) [](http://example.com/codeflow?ancestors=1936543215)

File: packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts [hunk 29]
```diff
@@ -170,7 +169,7 @@ describe("DDS Handle Encoding", () => {
 	];
 
 	testCases.forEach((testCase) => {
-		const shouldOrShouldNot = testCase.expectedHandles.length > 0 ? "should" : "should not";
+		const shouldOrShouldNot = testCase.expectedHandles.length > 0 ? "should not" : "should";
```
Comment (markfields) [inline packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts:172]: lol oops my bad!

File: packages/dds/ordered-collection/src/consensusOrderedCollection.ts [hunk 30]
```diff
@@ -140,6 +141,7 @@ export class ConsensusOrderedCollection<T = any>
 		await this.submit<IConsensusOrderedCollectionAddOperation<T>>({
 			opName: "add",
 			value: valueSer,
```
Comment (markfields) [inline packages/dds/ordered-collection/src/consensusOrderedCollection.ts:143]: Open an ADO task to stop writing this in a release or two?

Comment (connorskees) [reply packages/dds/ordered-collection/src/consensusOrderedCollection.ts:143]: Probably falls under https://dev.azure.com/fluidframework/internal/_workitems/edit/7149/

File: packages/dds/map/src/mapKernel.ts (hunk 20 above)
Comment (connorskees) [reply packages/dds/map/src/mapKernel.ts:185]: That `makeSerializable` is actually load-bearing and without it several hundred tests fail

File: experimental/dds/attributable-map/src/localValues.ts (hunk 28 above)
Comment (markfields) [reply experimental/dds/attributable-map/src/localValues.ts:137]: Is this still coming?

File: packages/dds/merge-tree/src/snapshotLoader.ts [hunk 31]
```diff
@@ -306,7 +306,10 @@ export class SnapshotLoader {
 	 */
 	private async loadCatchupOps(
 		rawMessages: Promise<ArrayBufferLike>,
+		serializer: IFluidSerializer,
 	): Promise<ISequencedDocumentMessage[]> {
-		return JSON.parse(bufferToString(await rawMessages, "utf8")) as ISequencedDocumentMessage[];
+		return serializer.parse(
```
Comment (markfields) [inline packages/dds/merge-tree/src/snapshotLoader.ts:311]: nit: Update `@returns` comment (it says "but handles aren't parsed")

File: packages/dds/sequence/src/defaultMap.ts [hunk 32]
```diff
@@ -262,9 +257,12 @@ export class DefaultMap<T extends IntervalCollection<any> = IntervalCollection<a
 
 	/**
 	 * Populate the kernel with the given map data.
-	 * @param data - A JSON string containing serialized map data
+	 *
+	 * @param serialized - A JSON string containing serialized map data
 	 */
-	public populateFromSerializable(json: IMapDataObjectSerializable): void {
+	public populate(serialized: string): void {
+		const json = this.serializer.parse(serialized) as IMapDataObjectSerializable;
```
Comment (markfields) [inline packages/dds/sequence/src/defaultMap.ts:264]: Let's call this somethign else... it has real handles in it... that's not JSON-like.

File: packages/dds/sequence/src/test/compare.py [hunk 33]
```diff
@@ -0,0 +1,46 @@
+import json
```
Comment (markfields) [inline packages/dds/sequence/src/test/compare.py:1]: Did you mean to include these scripts?

Comment (connorskees) [reply packages/dds/sequence/src/test/compare.py:1]: Ah must've slipped in during a merge conflict. Good catch!

File: packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts
[the remark this answers is not in the document]
Comment (markfields) [reply packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:1]: Inlcuding packages/test/functional-tests/src/test/ddsHandleEncoding.spec.ts

File: packages/dds/sequence/src/sequence.ts (hunk 11 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/sequence.ts:483]: loadedDeferredOutgoingOps is unused. it should be removed

File: packages/dds/map/src/mapKernel.ts (hunk 20 above)
Comment (markfields) [reply packages/dds/map/src/mapKernel.ts:185]: That makes me kind of worried.  Taking a quick look at the test failures.

**Review by markfields — APPROVED** [review]

**Review by markfields — APPROVED** [review]
Final updates look good, glad to have that extra serialization cleared up in Map.

File: packages/dds/shared-object-base/src/utils.ts [hunk 34]
```diff
@@ -78,3 +78,16 @@ export function createSingleBlobSummary(
 	builder.addBlob(key, content);
 	return builder.getSummaryTree();
 }
+
+/**
+ * Binds all handles found in `value` to `bind`. Does not modify original input.
```
Comment (agarwal-navin) [inline packages/dds/shared-object-base/src/utils.ts:83]:
```suggestion
 * Binds all IFluidHandles found in `value` to `bind`. Does not modify original input.
```

**Comment by agarwal-navin** [comment]
@connorskees @markfields This is a risky change. We should ensure that are enough tests in place to not break anything. The scenario where a combination of new and old runtime version clients are running in parallel should be validated. We should add basic end-to-end tests where there are 2 containers - one writes and other reads and run it in full compat mode so it does the mutli-client version compat testing.

**Comment by markfields** [comment]
@agarwal-navin -- I agree it is risky. Without any type validation to distinguish between encoded v. decoded handles, it's nearly impossible to tell whether all the flips/flops between those states have been addressed correctly.  In fact, I know of one still in the PR that hasn't, I've logged a follow-up item.

The reason I'm not overly concerned is that functions like `makeHandlesSerializable` and `parseHandles` are idempotent - in terms of correctness, all that matters is that on the way out, the handles are encoded _somewhere_ and on the way in it's the inverse.  Seeing that we have moved coverage here to the SharedObject choke points, I feel confident that _outside the DDS layer_ there's not going to be any behavior change here.  (inside the DDS layer is another story... counting on existing test coverage here)

So as for this:

> The scenario where a combination of new and old runtime version clients are running in parallel should be validated

I think we have a pretty good argument that there's no change over the wire.

That said, better safe than sorry, so the test you suggested is a good idea.  I wonder if we can find an existing test that is doing this incidentally?

@connorskees what do you think?

**Comment by markfields** [comment]
(Disabling auto-merge until @connorskees has a chance to respond to @agarwal-navin)

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/dds/shared-object-base/src/sharedObject.ts:474]: this shouldn't matter. the content is the previously submitted content, so it is already parsed. the dds then resubmits anything they want to be submitted which will subsequently get parses in the submit path again

File: packages/dds/sequence/src/sequence.ts (hunk 11 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/sequence.ts:483]: kind of. there is a flag that enables returning after loading the header, its not used, only in that case should loadedDeferredOutgoingOps have any ops in it.

**Review by anthony-murphy — APPROVED** [review]

**Review by Abe27342 — APPROVED** [review]
I looked at the tree changes only in latest revision. they look good. We could do further cleanup in experimental tree if we wanted to since the serializer/deserializer abstraction gets plumbed to more layers, but totally fine to push that out of this PR as is done.
